cpp/coord/fixed_point.h

```diff
@@ -0,0 +1,286 @@
+// Copyright 2013-2015 the openage authors. See copying.md for legal info.
+
+#ifndef OPENAGE_COORD_FIXED_POINT_H_
+#define OPENAGE_COORD_FIXED_POINT_H_
+
+#include <cmath>
+#include <functional>
+
+#include "../util/misc.h"
+
+namespace openage {
+namespace coord {
+
+template<typename int_type, unsigned fractional_bits>
+struct FixedPoint {
+
+	int_type data;
+
+	static const int_type one = int_type(1) << fractional_bits;
+
+	static const int_type fraction_bitmask = one - 1;
+
+
+	// constructors
+	constexpr FixedPoint() : data(0) {
+	}
+
+	constexpr FixedPoint(int_type n) : data(n << fractional_bits) {
+	}
+
+	constexpr FixedPoint(int n) : data(n << fractional_bits) {
+	}
+
+	constexpr FixedPoint(unsigned int n) : data(n << fractional_bits) {
+	}
+
+	// Constructing FixedPoint numbers with values like 1/2 and 1/8 won't lose precision
+	constexpr FixedPoint(float n) : data(static_cast<int_type>(n * one)) {
+	}
+
+	constexpr FixedPoint(double n) : data(static_cast<int_type>(n * one)) {
+	}
+
+	static FixedPoint from_raw(int_type data) {
+		FixedPoint result;
+		result.data = data;
+		return result;
+	}
+
+	constexpr FixedPoint(const FixedPoint &o) : data(o.data) {
+	}
+
+	FixedPoint &operator=(const FixedPoint &o) {
+		data = o.data;
+		return *this;
+	}
+
+
+	// Comparison operators
+	bool operator==(const FixedPoint &o) const {
```

Comment:
But then the compiler can optimize it away anyway as it's a `const &`. I'd say this is good the `&`-way.